README.md
# <img src="https://raw.githubusercontent.com/oognuyh/vscode-android-emulator-helper/master/images/logo.png" width="48" height="48" /> Android Emulator Helper [![Version](https://vsmarketplacebadge.apphb.com/version/oognuyh.android-emulator-helper.svg)](https://marketplace.visualstudio.com/items?itemName=oognuyh.android-emulator-helper) [![Installs](https://vsmarketplacebadge.apphb.com/installs-short/oognuyh.android-emulator-helper.svg)](https://marketplace.visualstudio.com/items?itemName=oognuyh.android-emulator-helper)

A Visual Studio Code extension that makes it easy to manage Android emulators.

## Features

| Create AVD | Delete AVD | Run Emulator |
|:---:|:---:|:---:|
| ![](https://raw.githubusercontent.com/oognuyh/vscode-android-emulator-helper/master/images/create-avd.gif) | ![](https://raw.githubusercontent.com/oognuyh/vscode-android-emulator-helper/master/images/delete-avd.gif) | ![](https://raw.githubusercontent.com/oognuyh/vscode-android-emulator-helper/master/images/run-emulator.gif) |

| Install Packages | Uninstall Packages | Update Installed Packages |
|:---:|:---:|:---:|
|![](https://raw.githubusercontent.com/oognuyh/vscode-android-emulator-helper/master/images/install-packages.gif) |![](https://raw.githubusercontent.com/oognuyh/vscode-android-emulator-helper/master/images/uninstall-packages.gif) | ![](https://raw.githubusercontent.com/oognuyh/vscode-android-emulator-helper/master/images/update-installed-packages.gif) |


> Note: If you encounter problems with SDK licenses, you can accept by clicking the ok button in the message. Please try again after this.

## Quick Start

- Install [Android Emulator Helper](https://marketplace.visualstudio.com/items?itemName=oognuyh.android-emulator-helper) extension.
- Open Cammand Palette (`Ctrl + Shift + p`).
- type Android Emulator Helper and choose a command.
    - Android Emulator Helper: Run Emulator
    - Android Emulator Helper: Create New AVD
    - Android Emulator Helper: Delete AVD
    - Android Emulator Helper: Install Packages
    - Android Emulator Helper: Uninstall Packages
    - Android Emulator Helper: Update All Installed Packages

## Settings Options

- android-emulator-helper.emulator-run-opts

    You can run your emulator with additional options. such as removing boot animation or resolution.
    Please refer to following page: [emulator options](https://android-doc.github.io/tools/help/emulator.html)


- android-emulator-helper.avd-creation-opts

    You can create your AVD with additional options. such as setting up a specific device.
    Please refer to following page: [avdmanager options](https://android-doc.github.io/tools/devices/managing-avds-cmdline.html)

- Example
    ```
    "android-emulator-helper": {
        "emulator-run-opts": "-no-boot-anim -wipe-data",
        "avd-creation-opts": "-d 30"
    }
    ```

## Requirements

- Android SDK
- Enviroment PATH variable
    - cmdline-tools/latest/bin or tools/bin
    - emulator
    - platform-tools


CHANGELOG.md
# Change Log

All notable changes to the "android-emulator-helper" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## 0.0.1

- Initial release of Android Emulator Helper

## 0.0.2

- Fix `--channel=0` command not found in Windows 10.
- Add progress bar when loading AVD or package list.

src/utils.ts
import { exec } from "child_process";
import { ProgressLocation, window } from "vscode";

interface CommandWithProgressProps {
  command: string;
  message: string;
  successMessage?: string;
  failureMessage?: string;
}

export const commandWithProgress = async ({
  command,
  message,
  successMessage,
  failureMessage,
}: CommandWithProgressProps): Promise<string> => {
  return window.withProgress(
    {
      location: ProgressLocation.Notification,
    },
    async (progress) =>
      new Promise((resolve, reject) => {
        const interval = setInterval(() => {
          progress.report({
            message: message,
          });
        });

        exec(command, (error, stdout, stderr) => {
          if (error) {
            console.error(stderr);
            window.showErrorMessage(
              failureMessage || `Failed to command: ${command}`
            );
            reject();
          }

          clearInterval(interval);

          progress.report({
            message: successMessage,
          });

          setTimeout(
            () => {
              resolve(stdout);
            },
            successMessage ? 1500 : 0
          );
        });
      })
  );
};

export const showAutocloseMessage = (message: string, s: number = 3) => {
  window.withProgress(
    {
      location: ProgressLocation.Notification,
    },
    async (progress) => {
      await new Promise((resolve) => {
        const interval = setInterval(() => {
          progress.report({
            message: message,
            increment: 100,
          });
        }, 500);

        setTimeout(() => {
          clearInterval(interval);
          resolve(undefined);
        }, s * 1000);
      });
    }
  );
};


src/extension.ts
import * as vscode from "vscode";
import {
  installPackages,
  uninstallPackages,
  updateAllInstalledPackages,
} from "./cmd/sdkmanager";
import { createAvd, deleteAvd } from "./cmd/avdmanager";
import { runEmulator } from "./cmd/emulator";

export function activate(context: vscode.ExtensionContext) {
  context.subscriptions.push(
    vscode.commands.registerCommand(
      "android-emulator-helper.install-packages",
      () => {
        installPackages();
      }
    )
  );

  context.subscriptions.push(
    vscode.commands.registerCommand(
      "android-emulator-helper.uninstall-packages",
      () => {
        uninstallPackages();
      }
    )
  );

  context.subscriptions.push(
    vscode.commands.registerCommand(
      "android-emulator-helper.create-avd",
      () => {
        createAvd();
      }
    )
  );

  context.subscriptions.push(
    vscode.commands.registerCommand(
      "android-emulator-helper.delete-avd",
      () => {
        deleteAvd();
      }
    )
  );

  context.subscriptions.push(
    vscode.commands.registerCommand(
      "android-emulator-helper.run-emulator",
      () => {
        runEmulator();
      }
    )
  );

  context.subscriptions.push(
    vscode.commands.registerCommand(
      "android-emulator-helper.update-all-installed-packages",
      () => {
        updateAllInstalledPackages();
      }
    )
  );
}

export function deactivate() {}


src/model/avd.ts
import { QuickPickItem } from "vscode";
import { Package } from "./package";

export class Avd implements QuickPickItem {
  label: string;
  description: string;
  detail: string;

  name: string;
  device: string;
  basedOn: string;
  package: Package | undefined;

  constructor(name: string, device: string, basedOn: string) {
    this.label = this.name = name;
    this.description = this.device = device;
    this.basedOn = this.detail = basedOn;
  }

  static from(rawAvd: string): Avd {
    const name: string =
      rawAvd
        .match(/name: (.)+/i)
        ?.shift()
        ?.replace(/name: /i, "") || "";
    const device: string =
      rawAvd
        .match(/device: (.)+/gi)
        ?.shift()
        ?.replace(/device: /i, "") || "";
    const basedOn: string =
      rawAvd
        .match(/based on: (.)+Tag\/ABI/i)
        ?.shift()
        ?.replace(/based on: |tag\/abi/gi, "") || "";

    return new Avd(name, device, basedOn);
  }
}


src/cmd/emulator.ts
import { window, workspace } from "vscode";
import { Avd } from "../model/avd";
import { commandWithProgress, showAutocloseMessage } from "../utils";
import { getAvailableAvds } from "./avdmanager";

export async function runEmulator() {
  const availableAvds: Avd[] | undefined = await getAvailableAvds();
  if (!availableAvds || availableAvds.length < 1) {
    showAutocloseMessage("You have no available AVD. Please create AVD first.");
    return;
  }

  const selectedAvd: Avd | undefined = await window.showQuickPick(
    availableAvds,
    {
      placeHolder: "Select AVD to run",
      canPickMany: false,
    }
  );
  if (!selectedAvd) {
    showAutocloseMessage("No AVD selected.");
    return;
  }

  const additonalEmulatorRunOpts: string =
    workspace
      .getConfiguration("android-emulator-helper")
      .get("emulator-run-opts") || "";

  const command: string = `emulator @${selectedAvd.name} ${additonalEmulatorRunOpts} > /dev/null 2>&1 &`;
  await commandWithProgress({
    command: command,
    message: `${selectedAvd.name} is being loaded...`,
    successMessage: `${selectedAvd.name} is successfully loaded.`,
    failureMessage: `Failed to load ${selectedAvd.name} with the command: ${command}.`,
  });
}


src/model/package.ts
import { QuickPickItem } from "vscode";

export class Package implements QuickPickItem {
  label: string;
  description: string;
  detail: string | undefined;

  isInstalled: boolean;
  path: string;
  version: string;

  constructor(
    path: string,
    version: string,
    description: string,
    isInstalled: boolean
  ) {
    this.path = this.label = path;
    this.version = version;
    this.description = description;
    this.isInstalled = isInstalled;

    if (this.isInstalled) {
      this.detail = "installed";
    }
  }

  static from(rawPackage: string, rawInstalledPackages: string): Package {
    const [path, version, description, ...etc]: string[] = rawPackage
      .split("|")
      .map((col) => col.trim());

    return new Package(
      path,
      version,
      description,
      rawInstalledPackages.includes(path)
    );
  }
}


src/cmd/avdmanager.ts
import { window, workspace } from "vscode";
import { Avd } from "../model/avd";
import { Package } from "../model/package";
import { commandWithProgress, showAutocloseMessage } from "../utils";
import { getInstalledPackages } from "./sdkmanager";

export async function createAvd() {
  // get installed pacakges
  const installedPackages = await getInstalledPackages("system-images;");
  if (!installedPackages || installedPackages.length < 1) {
    showAutocloseMessage(
      "No package available. You have to install the package first."
    );
    return;
  }

  // get package
  const selectedPackage: Package | undefined = await window.showQuickPick(
    installedPackages,
    {
      placeHolder: "Select a system image",
      canPickMany: false,
    }
  );
  if (!selectedPackage) {
    showAutocloseMessage("Can not create avd without package.");
    return;
  }

  const availableAvds = (await getAvailableAvds()) || [];

  // get new AVD name
  const newAvdName = await window.showInputBox({
    placeHolder: "Enter a new AVD name. (Must be unique)",
    validateInput: (name) => {
      if (name.match(/[^a-zA-Z0-9_]/)) {
        return `${name} is invalid! Must be [a-zA-Z0-9_]`;
      } else if (name.trim() === "") {
        return "Can't be blank!";
      } else if (
        availableAvds.filter((availableAvd) => availableAvd.name === name)
          .length > 0
      ) {
        return `${name} already exits!`;
      } else {
        return null;
      }
    },
  });
  if (!newAvdName) {
    showAutocloseMessage("Can not create AVD without name.");
    return;
  }

  const additionalAvdCreationOpts: string =
    workspace
      .getConfiguration("android-emulator-helper")
      .get("avd-creation-opts") || "";

  const command: string = `echo no | avdmanager create avd -n ${newAvdName} -k "${selectedPackage.path}" -f ${additionalAvdCreationOpts}`;
  await commandWithProgress({
    command: command,
    message: `${newAvdName} is being created...`,
    successMessage: `${newAvdName} created successfully.`,
    failureMessage: `Failed to create ${newAvdName} with the command: ${command}.`,
  });
}

export async function deleteAvd() {
  // get available AVDs
  const availableAvds: Avd[] | undefined = await getAvailableAvds();
  if (!availableAvds || availableAvds.length < 1) {
    showAutocloseMessage("You have no available AVD. Please create AVD first.");
    return;
  }

  // get AVD
  const selectedAvd: Avd | undefined = await window.showQuickPick(
    availableAvds,
    {
      placeHolder: "Select AVD name",
      canPickMany: false,
    }
  );
  if (!selectedAvd) {
    showAutocloseMessage("No AVD selected.");
    return;
  }

  const command = `avdmanager delete avd -n ${selectedAvd.name}`;
  await commandWithProgress({
    command: command,
    message: `${selectedAvd.name} is being deleted...`,
    successMessage: `${selectedAvd.name} deleted successfully.`,
    failureMessage: `Failed to delete ${selectedAvd.name} with the command: ${command}.`,
  });
}

export async function getAvailableAvds(): Promise<Avd[] | undefined> {
  const stdout: string = await commandWithProgress({
    command: "avdmanager list avd",
    message: "Loading available AVDs...",
  });

  const availableAvds: Avd[] | undefined = stdout
    .replace("Available Android Virtual Devices:\n", "")
    .split("---------\n")
    .map((rawAvailableAvd) => Avd.from(rawAvailableAvd));

  return availableAvds;
}


src/cmd/sdkmanager.ts
import { exec, execSync } from "child_process";
import { ProgressLocation, window } from "vscode";
import { Package } from "../model/package";
import { commandWithProgress, showAutocloseMessage } from "../utils";

export async function getAvailablePackages(): Promise<Package[] | undefined> {
  const stdout: string = await commandWithProgress({
    command: `sdkmanager --list`,
    message: `Loading available packages...`,
  });

  // parse
  const [rawInstalledPackages, rawPackages] = stdout.split(
    "Available Packages:\n"
  );

  return [...rawPackages.matchAll(/.*\|.*\|.*/g)]
    .slice(2)
    .map((rawPackage) => rawPackage[0])
    .map((rawPackage) => Package.from(rawPackage, rawInstalledPackages));
}

export async function installPackages() {
  const availablePackages: Package[] | undefined = await getAvailablePackages();
  if (!availablePackages || availablePackages.length < 1) {
    showAutocloseMessage(
      "No package available. You have to install the package first."
    );
    return;
  }

  const selectedPackages: Package[] | undefined = await window.showQuickPick(
    availablePackages,
    {
      placeHolder: "Select packages you want to install",
      canPickMany: true,
    }
  );
  if (!selectedPackages || selectedPackages.length < 1) {
    showAutocloseMessage("No pacakge selected.");
    return;
  }

  await window.withProgress(
    {
      location: ProgressLocation.Notification,
    },
    async (progress) => {
      let numOfSelectedPackages: number = selectedPackages.length;
      let numOfInstalledPackages: number = selectedPackages.length;
      let isProcessing: boolean = false;

      await new Promise((resolve, reject) => {
        let interval = setInterval(() => {
          if (!isProcessing) {
            if (selectedPackages.length > 0) {
              const selectedPackage: Package = selectedPackages.shift()!;

              isProcessing = true;
              progress.report({
                message: `${selectedPackage.path} is being installed.`,
              });

              // sdkmanager "something" "something" "something"
              exec(
                `sdkmanager "${selectedPackage.path}"`,
                (error, stdout, stderr) => {
                  if (error) {
                    console.error(stderr);

                    if (stderr.includes("license")) {
                      // if user need to accept the licenses
                      clearInterval(interval);
                      reject();

                      window
                        .showErrorMessage(
                          "You need to accept the licenses before building. You want to accept now?",
                          ...["OK", "Later"]
                        )
                        .then((selection) => {
                          if (selection === "OK") {
                            const result = execSync(
                              "echo yes | sdkmanager --licenses"
                            ).toString();

                            if (result.includes("accepted")) {
                              window.showInformationMessage(
                                "All SDK package licenses accepted. Please try to install again."
                              );
                            } else {
                              window.showErrorMessage(
                                "Can't accept the licenses. Please try manually."
                              );
                            }
                          }
                        });
                    } else {
                      window.showErrorMessage(
                        `Failed to install ${selectedPackage.path}`
                      );

                      numOfInstalledPackages--;
                      isProcessing = false;
                      return;
                    }
                  }

                  // installed successfully
                  progress.report({
                    increment: (1 / numOfSelectedPackages) * 100,
                  });

                  isProcessing = false;
                }
              );
            } else {
              // done
              clearInterval(interval);
              progress.report({
                message: `${numOfInstalledPackages} packages installed successfully.`,
              });

              setTimeout(() => {
                resolve(undefined);
              }, 1500);
            }
          }
        }, 500);
      });
    }
  );
}

export async function getInstalledPackages(
  filter: string | undefined = undefined
): Promise<Package[] | undefined> {
  // load installed packages
  const stdout: string = await commandWithProgress({
    command: `sdkmanager --list_installed`,
    message: `Loading installed packages...`,
  });

  // parse
  let installedPackages: Package[] | undefined = [
    ...stdout.matchAll(/.*\|.*\|.*\|.*\n/g),
  ]
    .map((rawInstalledPackage) => rawInstalledPackage[0])
    .slice(2)
    .map((rawInstalledPackage) => Package.from(rawInstalledPackage, stdout));

  if (filter) {
    installedPackages = installedPackages.filter((installPackage) =>
      installPackage.path.includes(filter)
    );
  }

  return installedPackages;
}

export async function uninstallPackages() {
  const installedPackages: Package[] | undefined = await getInstalledPackages();
  if (!installedPackages || installedPackages.length < 1) {
    showAutocloseMessage("You have no package.");
    return;
  }

  const selectedPackages: Package[] | undefined = await window.showQuickPick(
    installedPackages,
    {
      placeHolder: "Select packages to uninstall",
      canPickMany: true,
    }
  );
  if (!selectedPackages || selectedPackages.length < 1) {
    showAutocloseMessage("No package selected.");
    return;
  }

  await window.withProgress(
    {
      location: ProgressLocation.Notification,
    },
    async (progress) => {
      let numOfSelectedPackages: number = selectedPackages.length;
      let numOfUninstalledPackages: number = selectedPackages.length;
      let isProcessing: boolean = false;

      await new Promise((resolve) => {
        let interval = setInterval(() => {
          if (!isProcessing) {
            if (selectedPackages.length > 0) {
              const selectedPackage: Package = selectedPackages.shift()!;

              isProcessing = true;
              progress.report({
                message: `${selectedPackage.path} is being uninstalled.`,
              });

              // sdkmanager --uninstall "something" "something" "something"
              exec(
                `sdkmanager --uninstall "${selectedPackage.path}"`,
                (error, stdout, stderr) => {
                  if (error) {
                    window.showErrorMessage(
                      `Failed to uninstall ${selectedPackage.path}`
                    );

                    console.error(stdout);

                    numOfUninstalledPackages--;
                    isProcessing = false;
                    return;
                  }

                  // installed successfully
                  progress.report({
                    increment: (1 / numOfSelectedPackages) * 100,
                  });
                  isProcessing = false;
                }
              );
            } else {
              // done
              progress.report({
                message: `${numOfUninstalledPackages} packages uninstalled successfully.`,
              });

              setTimeout(() => {
                clearInterval(interval);
                resolve(undefined);
              }, 1500);
            }
          }
        }, 500);
      });
    }
  );
}

export async function updateAllInstalledPackages() {
  await window.withProgress(
    {
      location: ProgressLocation.Notification,
    },
    async (progress) => {
      await new Promise((resolve, reject) => {
        const interval = setInterval(() => {
          progress.report({
            message: "All installed packages are being updated.",
          });
        }, 500);

        exec(
          "echo yes | sdkmanager --update --channel=0",
          (error, stdout, stderr) => {
            if (error) {
              console.error(stderr);

              if (stderr.includes("license")) {
                // if user need to accept the licenses
                clearInterval(interval);
                reject();

                window
                  .showErrorMessage(
                    "You need to accept the licenses before building. You want to accept now?",
                    ...["OK", "Later"]
                  )
                  .then((selection) => {
                    if (selection === "OK") {
                      const result = execSync(
                        "echo yes | sdkmanager --licenses"
                      ).toString();

                      if (result.includes("accepted")) {
                        window.showInformationMessage(
                          "All SDK package licenses accepted. Please try to update again."
                        );
                      } else {
                        window.showErrorMessage(
                          "Can't accept the licenses. Please try manually."
                        );
                      }
                    }
                  });
              } else {
                window.showErrorMessage(`Failed to update the packages.`);
              }
            }

            clearInterval(interval);
            progress.report({
              message: stdout.includes("No updates available")
                ? "No updates available."
                : "All pacakges updated successfully.",
            });

            setTimeout(() => {
              resolve(undefined);
            }, 2000);
          }
        );
      });
    }
  );
}
